File: Assignment 5/C/09. reverse a doubly linked list.c
// tips: just change the value in a reverse manner 
#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    struct node* prev;
    int val;
    struct node* next;
}node;

void traversal(node* head){
    node* ptr = head;
    while(ptr){
        printf("%d ",ptr->val);
        ptr = ptr->next;
    }
    printf("\n");
}

void reverse(node* head){
    int s = 0;
    node* ptr = head;
    while(ptr->next){
        s++;
        ptr = ptr->next;
    }
    node* str = head;
    int i=0;
    int j=s+1;
    while(i<j){
        int t = ptr->val;
        ptr->val = str->val;
        str->val = t;
        str = str->next;
        ptr = ptr->prev;
        i++;
        j--;
    }
}

int main(){
    node* head = (node*)malloc(sizeof(node));
    node* snd = (node*)malloc(sizeof(node));
    node* trd = (node*)malloc(sizeof(node));
    node* four = (node*)malloc(sizeof(node));

    head->val = 1;
    snd->val = 2;
    trd->val = 3;
    four->val = 4;

    head->prev = NULL;
    head->next = snd;

    snd->prev = head;
    snd->next = trd;
    
    trd->prev = snd;
    trd->next = four;

    four->prev = trd;
    four->next = NULL;

    printf("before: ");
    traversal(head);

    printf("after: ");
    reverse(head);
    traversal(head);

    return 0;
}

File: Assignment 5/Python/09. reverse a doubly linked list .py
class Node:
    def __init__(self, val):
        self.prev = None
        self.val = val
        self.next = None

def traversal(head):
    ptr = head
    while ptr:
        print(ptr.val, end=" ")
        ptr = ptr.next
    print()

def reverse(head):
    length = 0
    ptr = head
    while ptr.next:
        length += 1
        ptr = ptr.next

    str_ptr = head
    i = 0
    j = length + 1
    while i < j:
        temp = ptr.val
        ptr.val = str_ptr.val
        str_ptr.val = temp
        str_ptr = str_ptr.next
        ptr = ptr.prev
        i += 1
        j -= 1

if __name__ == "__main__":
    head = Node(1)
    snd = Node(2)
    trd = Node(3)
    four = Node(4)

    head.prev = None
    head.next = snd

    snd.prev = head
    snd.next = trd

    trd.prev = snd
    trd.next = four

    four.prev = trd
    four.next = None

    print("Before: ", end="")
    traversal(head)
    reverse(head)
    print("After: ", end="")
    traversal(head)

# Output:
# Before: 1 2 3 4 
# After: 4 3 2 1 
